File: 500.600/538.ConvertBstToGreaterTree.cpp
#include <iostream>
using namespace std;

struct TreeNode
{
    int val{};
    TreeNode *left{};
    TreeNode *right{};
    TreeNode() {}
    TreeNode(int x) : val(x) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class ConvertBstToGreaterTree
{
public:
    TreeNode *convertBST(TreeNode *root)
    {
        if (!root)
        {
            return root;
        }
        num = 0;
        dfs(root);
        return root;
    }

private:
    void dfs(TreeNode *node)
    {
        if (node->right)
        {
            dfs(node->right);
        }
        num += node->val;
        node->val = num;
        if (node->left)
        {
            dfs(node->left);
        }
    }

private:
    int num{};

public:
    TreeNode *convertBST1(TreeNode *root)
    {
        int sum = 0;
        TreeNode *node = root;
        while (node != nullptr)
        {
            if (node->right == nullptr)
            {
                sum += node->val;
                node->val = sum;
                node = node->left;
            }
            else
            {
                TreeNode *succ = getSuccessor(node);
                if (succ->left == nullptr)
                {
                    // 建立新的临时路径
                    succ->left = node;
                    node = node->right;
                }
                else
                {
                    // 删除新建的临时路径
                    succ->left = nullptr;
                    sum += node->val;
                    node->val = sum;
                    node = node->left;
                }
            }
        }
        return root;
    }

private:
    TreeNode *getSuccessor(TreeNode *node)
    {
        TreeNode *succ = node->right;
        while (succ->left != nullptr && succ->left != node)
        {
            succ = succ->left;
        }

        return succ;
    }
};

void printAndDel(TreeNode *node)
{
    if (node->left)
    {
        printAndDel(node->left);
    }
    cout << node->val << ",";
    if (node->right)
    {
        printAndDel(node->right);
    }
    delete node;
}

int main(int argc, char const *argv[])
{
    TreeNode *head = new TreeNode(4);
    head->left = new TreeNode(1);
    head->left->left = new TreeNode(0);
    head->left->right = new TreeNode(2);
    head->left->right->right = new TreeNode(3);
    head->right = new TreeNode(6);
    head->right->left = new TreeNode(5);
    head->right->right = new TreeNode(7);
    head->right->right->right = new TreeNode(8);
    ConvertBstToGreaterTree cls;
    printAndDel(cls.convertBST1(head));
    cout << endl;

    head = new TreeNode(0);
    head->right = new TreeNode(1);
    printAndDel(cls.convertBST(head));
    cout << endl;

    head = new TreeNode(1);
    head->left = new TreeNode(0);
    head->right = new TreeNode(2);
    printAndDel(cls.convertBST(head));
    cout << endl;

    head = new TreeNode(3);
    head->left = new TreeNode(2);
    head->left->left = new TreeNode(1);
    head->right = new TreeNode(4);
    printAndDel(cls.convertBST(head));
    cout << endl;

    return 0;
}

File: go/500.600/538.ConvertBstToGreaterTree.go
package main

import "fmt"

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

var sum = 0

func createTreeNode(v int) *TreeNode {
	return &TreeNode{Val: v}
}

func convertBST(root *TreeNode) *TreeNode {
	sum = 0
	bfs(root)
	return root
}

func bfs(root *TreeNode) {
	if root == nil {
		return
	}
	bfs(root.Right)
	sum += root.Val
	root.Val = sum
	bfs(root.Left)
}

func print(root *TreeNode) {
	if root.Right != nil {
		print(root.Right)
	}
	fmt.Printf("%d,", root.Val)
	if root.Left != nil {
		print(root.Left)
	}
}

func convertBST1(root *TreeNode) *TreeNode {
	sum := 0
	node := root
	for node != nil {
		if node.Right == nil {
			sum += node.Val
			node.Val = sum
			node = node.Left
		} else {
			succ := getSucc(node)
			if succ.Left == nil {
				succ.Left = node
				node = node.Right
			} else {
				succ.Left = nil
				sum += node.Val
				node.Val = sum
				node = node.Left
			}
		}

	}
	return root
}

func getSucc(node *TreeNode) *TreeNode {
	succ := node.Right
	for succ.Left != nil && succ.Left != node {
		succ = succ.Left
	}
	return succ
}

func main() {
	root := createTreeNode(4)
	root.Left = createTreeNode(1)
	root.Left.Left = createTreeNode(0)
	root.Left.Right = createTreeNode(2)
	root.Left.Right.Right = createTreeNode(3)
	root.Right = createTreeNode(6)
	root.Right.Left = createTreeNode(5)
	root.Right.Right = createTreeNode(7)
	root.Right.Right.Right = createTreeNode(8)
	root = convertBST1(root)
	print(root)
	fmt.Println()
}
